instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/UpcomingRouteObjectsTest.kt

```diff
@@ -54,6 +67,80 @@ class UpcomingRouteObjectsTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::
+            }
+            .upcomingRoadObjects
+        assertEquals(7, upcomingRoadObjectsOnStart.size)
+        assertEquals(220.27365, upcomingRoadObjectsOnStart[0].distanceToStart!!, tolerance)
+        assertEquals(2421.0498, upcomingRoadObjectsOnStart[1].distanceToStart!!, tolerance)
+        assertEquals(1789.1295, upcomingRoadObjectsOnStart[2].distanceToStart!!, tolerance)
+        assertEquals(4576.7544, upcomingRoadObjectsOnStart[3].distanceToStart!!, tolerance)
+        assertEquals(4576.7544, upcomingRoadObjectsOnStart[4].distanceToStart!!, tolerance)
+        assertEquals(9232.6493, upcomingRoadObjectsOnStart[5].distanceToStart!!, tolerance)
+        assertEquals(9434.4270, upcomingRoadObjectsOnStart[6].distanceToStart!!, tolerance)
+
+        val distanceDiff = 1763.9070399
+        stayOnPosition(positionAlongTheRoute)
+        val upcomingRoadObjectsAlongTheRoute = mapboxNavigation.routeProgressUpdates()
+            .first {
+                it.currentState == RouteProgressState.TRACKING &&
+                    it.currentRouteGeometryIndex > 0
+            }
+            .upcomingRoadObjects
+        assertEquals(6, upcomingRoadObjectsAlongTheRoute.size)
+        upcomingRoadObjectsAlongTheRoute.forEachIndexed { index, objectAlongTheRoute ->
+            assertEquals(
+                "Object along the route #$index",
+                upcomingRoadObjectsOnStart[index + 1].distanceToStart!! - distanceDiff,
+                objectAlongTheRoute.distanceToStart!!,
+                tolerance
+            )
+        }
+    }
+
+    @Test
+    fun icAndJctTest() = sdkTest {
+        val jctOrigin = Point.fromLngLat(139.790845, 35.634688)
+        stayOnPosition(jctOrigin)
+        mapboxNavigation.startTripSession()
+        setUpRoutes(R.raw.route_with_jct, "139.790833,35.63468;139.778821,35.635955")
+
+        val upcomingJunctions = mapboxNavigation.routeProgressUpdates()
+            .first { it.currentState == RouteProgressState.TRACKING }
+            .upcomingRoadObjects
+            .filter { it.roadObject.objectType == RoadObjectType.JCT }
+        assertEquals(1, upcomingJunctions.size)
+        assertEquals("Ariake JCT", (upcomingJunctions[0].roadObject as Junction).name[0].value)
+
+        val icOrigin = Point.fromLngLat(140.025875, 35.66031)
+        stayOnPosition(icOrigin)
+        setUpRoutes(R.raw.route_with_ic, "140.025878,35.660315;140.019419,35.664076")
+
+        val upcomingInterchanges = mapboxNavigation.routeProgressUpdates()
+            .first {
+                it.currentState == RouteProgressState.TRACKING &&
+                    it.navigationRoute.id.startsWith("route_with_ic")
+            }
+            .upcomingRoadObjects
+            .filter { it.roadObject.objectType == RoadObjectType.IC }
+        assertEquals(1, upcomingInterchanges.size)
+        assertEquals(
+            "Wangannarashino IC",
+            (upcomingInterchanges[0].roadObject as Interchange).name[0].value
+        )
```

Comment:
Does this part have any relation to the beginning of the test? Does it make sense to move it to a separate tests?

Reply:
What exactly do you want to move to a separate test? There are 2 different routes, that's right. I could write 2 tests, I just wanted to make it one, because stopping/starting a test also takes up some time. Just a little optimization.

Reply:
> What exactly do you want to move to a separate test?

The selected lines.

> I could write 2 tests, I just wanted to make it one, because stopping/starting a test also takes up some time. Just a little optimization.

That's a fair point. That way tests work faster. Have you considered possibility of using `BaseCoreNoCleanUpTest`. Maybe 2 tests with with less presetup, which seems unnecessary for this specific case, will be faster than the current one? 🤔

Reply:
Hm, as far as I understand, I can use `BaseCoreNoCleanUpTest` almost anywhere. Will keep that in mind.